Add `java.jms` specification (#110)
ballerina-platform/module-ballerinax-java.jms: docs/spec/spec.md

# Specification: Ballerina `java.jms` Library

_Owners_: @shafreenAnfar @ayeshLK \
_Reviewers_: @shafreenAnfar \
_Created_: 2023/08/15 \
_Updated_: 2023/08/17 \
_Edition_: Swan Lake 

## Introduction  

This is the specification for the `java.jms` standard library of [Ballerina language](https://ballerina.io/), which provides the 
functionality to send and receive messages by connecting to a JMS provider. 

The `java.jms` library specification has evolved and may continue to evolve in the future. The released versions of the 

> specification can be found under the relevant Github tag.

Reviewer: ```suggestion
specification can be found under the relevant GitHub tag.
```

    * 2.1. [Configuration](#21-configuration)
    * 2.2. [Initialization](#22-initialization)
    * 2.3. [Functions](#23-functions)
3. [Session](#3-session)
    * 3.1. [Configuration](#31-configuration)
    * 3.2. [Functions](#32-functions)
4. [Message](#4-message)
5. [Message producer](#5-message-producer)
   * 5.1. [Functions](#51-functions)
6. [Message consumer](#6-message-consumer)
   * 6.1. [Functions](#61-functions)
7. [Message listener](#7-message-listener)
   * 7.1. [Configuration](#71-configuration)
   * 7.2. [Initialization](#72-initialization)
   * 7.3. [Functions](#73-functions)
   * 7.4. [Caller](#74-caller)
     * 7.4.1. [Functions](#741-functions)
   * 7.5. [Usage](#75-usage)

## 1. Overview  

Java Message Service (JMS) is a Java-based API that provides a standardized way for applications to create, send, 
receive, and consume messages in a loosely coupled, reliable, and asynchronous manner. JMS is part of the Java EE 
(Enterprise Edition) specification and is commonly used for building distributed, decoupled, and scalable applications 
that need to exchange information or events between different components. This specification elaborates on the usage of 

> JMS connection, session, message producer, and message consumer. These clients allow writing distributed applications 

Reviewer: ```suggestion
JMS connection, session, message producer, and message consumer. These clients allow the writing of distributed applications 
```

and microservices that read, write, and process messages using a JMS provider. 

Ballerina `java.jms` provides several core apis:
- `jms:Connection` - represents a communication link between JMS client and JMS provider.
- `jms:Session` - represents a single-threaded context for producing and consuming messages.
- `jms:MessageProducer` - used to send messages to a specific destination (queue or a topic) within the JMS session.
- `jms:MessageConsumer` - used to receive messages from a specific destination (queue or a topic) within the JMS session.
- `jms:Listener` - used to receive messages from a specific destination (queue or a topic) in an asynchronous manner.

## 2. Connection

A JMS connection represents a communication link between a JMS client and a JMS provider (typically a messaging broker). 
It's established using a JMS connection factory. The connection manages underlying resources, such as network 
connections, and provides a context for creating JMS sessions.

### 2.1. Configuration

> When initializing a `jms:Connection`, following configurations can be provided.

Reviewer: ```suggestion
When initializing a `jms:Connection`, the following configurations can be provided.
```

```ballerina
# Configurations related to a JMS connection.
#

> # + initialContextFactory - JMS provider specific inital context factory

Reviewer: ```suggestion
# + initialContextFactory - JMS provider-specific inital context factory
```

    string initialContextFactory;
    string providerUrl;
    string connectionFactoryName = "ConnectionFactory";
    string username?;
    string password?;
    map<string> properties = {};
|};
```

### 2.2. Initialization

The `jms:Connection` can be initialized by providing the `jms:ConnectionConfiguration`.
```ballerina
# Initialize and starts a JMS connection.
# ```
# jms:Connection connection = check new (
#   initialContextFactory = "org.apache.activemq.jndi.ActiveMQInitialContextFactory",
#   providerUrl = "tcp://localhost:61616"
# );
# ```
#
# + connectionConfig - The configurations to be used when initializing the JMS connection
# + return - The `jms:Connection` or an `jms:Error` if the initialization failed
public isolated function init(*jms:ConnectionConfiguration connectionConfig) returns jms:Error?;
```

### 2.3. Functions

To start (or restart) a connection's delivery of incoming messages for a connection, the `'start` function can be used.
```ballerina
# Starts (or restarts) a connection's delivery of incoming messages.
# A call to start on a connection that has already been started is ignored.
# ```
# check connection->'start()
# ```
#

> # + return - A `jms:Error` if threre is an error while starting the connection

Reviewer: ```suggestion
# + return - A `jms:Error` if there is an error while starting the connection
```

isolated remote function 'start() returns jms:Error?;
```

To temporarily stop the connection's delivery of incoming messages for a connection, the `stop` function can be used.
```ballerina
# Temporarily stops a connection's delivery of incoming messages.
# Delivery can be restarted using the connection's start method.
# ```
# check connection->stop();
# ```
#

> # + return - A `jms:Error` if threre is an error while stopping the connection

Reviewer: ```suggestion
# + return - A `jms:Error` if there is an error while stopping the connection
```

isolated remote function stop() returns jms:Error?;
```

To close the JMS connection, the `close` function can be used.
```ballerina
# Closes the connection.
# ```
# check connection->close();
# ```
#
# + return - A `jms:Error` if threre is an error while closing the connection
isolated remote function close() returns jms:Error?;
```

To initialize a new `jms:Session`, the `createSession` function can be used.
```ballerina

> # Defines the JMS session acknowledgement modes.

Reviewer: ```suggestion
# Defines the JMS session acknowledgment modes.
```

Author: This spellings is correct.

public enum AcknowledgementMode {
    # Indicates that the session will use a local transaction which may subsequently 
    # be committed or rolled back by calling the session's `commit` or `rollback` methods. 
    SESSION_TRANSACTED = "SESSION_TRANSACTED",
    # Indicates that the session automatically acknowledges a client's receipt of a message 
    # either when the session has successfully returned from a call to `receive` or when 
    # the message listener the session has called to process the message successfully returns.
    AUTO_ACKNOWLEDGE = "AUTO_ACKNOWLEDGE",
    # Indicates that the client acknowledges a consumed message by calling the 
    # MessageConsumer's or Caller's `acknowledge` method. Acknowledging a consumed message 
    # acknowledges all messages that the session has consumed.
    CLIENT_ACKNOWLEDGE = "CLIENT_ACKNOWLEDGE",

>     # Indicates that the session to lazily acknowledge the delivery of messages. 

Reviewer: ```suggestion
    # Indicates that the session lazily acknowledges the delivery of messages. 
```

}

# Message consumer listener configurations.
#
# + type - Message consumer type
# + destination - Name of the JMS destination
# + messageSelector - only messages with properties matching the message selector expression are added to the durable subscription. 
#                     An empty string indicates that there is no message selector for the durable subscription.
# + noLocal - if true then any messages published to the topic using this session's connection, or any other connection 
#             with the same client identifier, will not be added to the durable subscription.
# + subscriberName - the name used to identify the subscription 
public type ConsumerOptions record {|
    jms:ConsumerType 'type = DEFAULT;
    jms:Destination destination;
    string messageSelector = "";
    boolean noLocal = false;
    string subscriberName?;
|};
```

### 3.2. Functions

To unsubscribe a durable subscription that has been created by the JMS session, `unsubscribe` function can be used.
```ballerina
# Unsubscribe a durable subscription that has been created by this session.
# It is erroneous for a client to delete a durable subscription while there is an active (not closed) consumer

> # for the subscription, or while a consumed message being part of a pending transaction or has not been

Reviewer: ```suggestion
# for the subscription, or while a consumed message is part of a pending transaction or has not been
```

function can be used.
```ballerina
# Rolls back any messages sent/received in this transaction and releases any locks currently held.
# ```
# check session->'rollback();
# ```
# 
# + return - A `jms:Error` if there is an error or else `()`
isolated remote function 'rollback() returns jms:Error?;
```

To close the current JMS session, `close` function can be used.
```ballerina
# Closes the current JMS session.
# ```ballerina
# check session->close();
# ```
# 
# + return - A `jms:Error` if there is an error or else `()`
isolated remote function close() returns jms:Error?;
```

## 4. Message

A JMS message is a unit of data representing information that is exchanged between components using the Java Message 
Service (JMS) API. Messages can contain various types of data, such as text, binary data, maps, and objects. 
Ballerina `java.jms` library supports messages of type text, binary data, and maps.
```ballerina
# Represent the JMS Message used to send and receive content from the a JMS provider.
#
# + messageId - Unique identifier for a JMS message  
# + timestamp - Time a message was handed off to a provider to be sent 

> # + correlationId - Id which can be use to correlate multiple messages 

Reviewer: ```suggestion
# + correlationId - Id which can be used to correlate multiple messages 
```

# ```
# check producer->sendTo({ 'type: QUEUE, name: "test-queue" }, message);
# ```
#
# + destination - Destination used for the message sender
# + message - Message to be sent to the JMS provider
# + return - A `jms:Error` if there is an error or else `()`
isolated remote function sendTo(jms:Destination destination, jms:Message message) returns jms:Error?;
```

To close the message producer, `close` function can be used.
```ballerina
# Closes the message producer.
# ``` 
# check producer->close();
# ```
# + return - A `jms:Error` if there is an error or else `()`
isolated remote function close() returns jms:Error?;
```

## 6. Message consumer

A JMS message consumer is used to receive messages from a specific destination (queue or topic) within a JMS session. 
It's created from a JMS session and provides methods to receive messages synchronously.

### 6.1. Functions

To receive the next message arriving within the specified timeout interval, `receive` function can be used.
```ballerina
# Receives the next message that arrives within the specified timeout interval.
# ```
# jms:Message message = check consumer->receive(10000);
# ```
#
# + timeoutMillis - Message receive timeout

> # + return - A `jms:JmsMessage` if there is a new message, `()` if there is no any new message, 

Reviewer: ```suggestion
# + return - A `jms:JmsMessage` if there is a new message, `()` if there is no new message, 
```

# or else a `jsm:Error` if there is an error in the execution
isolated remote function receive(int timeoutMillis = 10000) returns jms:Message|jms:Error?;
```

To receive the next message if one is immediately available, `receiveNoWait` function can be used.
```ballerina
# Receives the next message if one is immediately available.
# ```
# jms:Message? message = check consumer->receiveNoWait();
# ```
# 
# + return - A `jms:JmsMessage` if there is a new message, `()` if there is no any new message, 
# or else a `jsm:Error` if there is an error in the execution
isolated remote function receiveNoWait() returns jms:Message|jms:Error?;
```

> To mark a JMS message as received, `acknowledge` function can be used.

Reviewer: ```suggestion
To mark a JMS message as received, the `acknowledge` function can be used.
```
